src/System.Net.WebProxy/src/System/Net/WebProxy.cs

```diff
@@ -0,0 +1,238 @@
+            }
+
+            return IsBypassed(destination) ? destination : Address;
+        }
+
+        private static Uri CreateProxyUri(string address) =>
+            address == null ? null :
+            address.IndexOf("://") == -1 ? new Uri("http://" + address) :
+            new Uri(address);
+
+        private void UpdateRegExList(bool canThrow)
+        {
+            Regex[] regExBypassList = null;
+            ArrayList bypassList = _bypassList;
+            try
+            {
+                if (bypassList != null && bypassList.Count > 0)
+                {
+                    regExBypassList = new Regex[bypassList.Count];
+                    for (int i = 0; i < bypassList.Count; i++)
+                    {
+                        regExBypassList[i] = new Regex((string)bypassList[i], RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
+                    }
+                }
+            }
+            catch
+            {
+                if (!canThrow)
+                {
+                    _regExBypassList = null;
+                    return;
+                }
+                throw;
+            }
+
+            // Update field here, as it could throw earlier in the loop
+            _regExBypassList = regExBypassList;
+        }
+
+        private bool IsMatchInBypassList(Uri input)
+        {
+            UpdateRegExList(false);
+
+            if (_regExBypassList != null)
+            {
+                string matchUriString = input.IsDefaultPort ?
+                    input.Scheme + "://" + input.Host :
+                    input.Scheme + "://" + input.Host + ":" + input.Port.ToString();
+
+                foreach (Regex r in _regExBypassList)
+                {
+                    if (r.IsMatch(matchUriString))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private bool IsLocal(Uri host)
+        {
+            string hostString = host.Host;
+
+            IPAddress hostAddress;
+            if (IPAddress.TryParse(hostString, out hostAddress))
+            {
+                return IPAddress.IsLoopback(hostAddress) || IsAddressLocal(hostAddress);
+            }
+
+            // No dot?  Local.
+            int dot = hostString.IndexOf('.');
+            if (dot == -1)
+            {
+                return true;
+            }
+
+            // If it matches the primary domain, it's local.  (Whether or not the hostname matches.)
+            string local = "." + IPGlobalProperties.GetIPGlobalProperties().DomainName;
+            return
+                local.Length == (hostString.Length - dot) &&
+                string.Compare(local, 0, hostString, dot, local.Length, StringComparison.OrdinalIgnoreCase) == 0;
+        }
+
+        private static bool IsAddressLocal(IPAddress ipAddress)
+        {
+            // TODO: The .NET Framework caches this and then uses network change
```

Comment:
I wanted feedback on this.  Do you think we should:
- Cache on first use?
- Cache but invalidate the cache when a network change event occurs (which means registering for it)?
- Not cache and just get the list every time?